mongo/integration/collection_test.go

```diff
@@ -484,6 +484,96 @@ func TestCollection(t *testing.T) {
+			assert.NotNil(mt, err, "expected error, got nil")
+		})
+		mt.Run("nil id", func(mt *mtest.T) {
+			_, err := mt.Coll.UpdateByID(mtest.Background, nil, bson.D{{"$inc", bson.D{{"x", 1}}}})
+			assert.NotNil(mt, err, "expected error, got nil")
+		})
+		mt.RunOpts("found", noClientOpts, func(mt *mtest.T) {
+			testCases := []struct {
+				name string
+				id   interface{}
+			}{
+				{"objectID", primitive.NewObjectID()},
+				{"string", "foo"},
+				{"int", 11},
+			}
+			for _, tc := range testCases {
+				mt.Run(tc.name, func(mt *mtest.T) {
+					doc := bson.D{{"_id", tc.id}, {"x", 1}}
+					_, err := mt.Coll.InsertOne(mtest.Background, doc)
+					assert.Nil(mt, err, "InsertOne error: %v", err)
+
+					update := bson.D{{"$inc", bson.D{{"x", 1}}}}
+
+					res, err := mt.Coll.UpdateByID(mtest.Background, tc.id, update)
+					assert.Nil(mt, err, "UpdateOne error: %v", err)
+					assert.Equal(mt, int64(1), res.MatchedCount, "expected matched count 1, got %v", res.MatchedCount)
+					assert.Equal(mt, int64(1), res.ModifiedCount, "expected matched count 1, got %v", res.ModifiedCount)
+					assert.Nil(mt, res.UpsertedID, "expected upserted ID nil, got %v", res.UpsertedID)
+				})
+			}
+		})
+		mt.Run("not found", func(mt *mtest.T) {
+			id := primitive.NewObjectID()
+			doc := bson.D{{"_id", id}, {"x", 1}}
+			_, err := mt.Coll.InsertOne(mtest.Background, doc)
+			assert.Nil(mt, err, "InsertOne error: %v", err)
+
+			update := bson.D{{"$inc", bson.D{{"x", 1}}}}
+
+			res, err := mt.Coll.UpdateByID(mtest.Background, 0, update)
+			assert.Nil(mt, err, "UpdateOne error: %v", err)
+			assert.Equal(mt, int64(0), res.MatchedCount, "expected matched count 0, got %v", res.MatchedCount)
+			assert.Equal(mt, int64(0), res.ModifiedCount, "expected matched count 0, got %v", res.ModifiedCount)
+			assert.Nil(mt, res.UpsertedID, "expected upserted ID nil, got %v", res.UpsertedID)
+		})
+		mt.Run("upsert", func(mt *mtest.T) {
+			id := primitive.NewObjectID()
+			doc := bson.D{{"_id", id}, {"x", 1}}
+			_, err := mt.Coll.InsertOne(mtest.Background, doc)
+			assert.Nil(mt, err, "InsertOne error: %v", err)
+
+			update := bson.D{{"$inc", bson.D{{"x", 1}}}}
+
+			res, err := mt.Coll.UpdateByID(mtest.Background, "blah", update, options.Update().SetUpsert(true))
+			assert.Nil(mt, err, "UpdateOne error: %v", err)
+			assert.Equal(mt, int64(0), res.MatchedCount, "expected matched count 0, got %v", res.MatchedCount)
+			assert.Equal(mt, int64(0), res.ModifiedCount, "expected matched count 0, got %v", res.ModifiedCount)
+			assert.NotNil(mt, res.UpsertedID, "expected upserted ID, got nil")
+		})
+		mt.Run("write error", func(mt *mtest.T) {
+			id := "foo"
+			update := bson.D{{"$set", bson.D{{"_id", 3.14159}}}}
+			_, err := mt.Coll.InsertOne(mtest.Background, bson.D{{"_id", id}})
+			assert.Nil(mt, err, "InsertOne error: %v", err)
+
+			_, err = mt.Coll.UpdateByID(mtest.Background, id, update)
+			we, ok := err.(mongo.WriteException)
```

Comment:
It depends on how important we consider checking for 1 writeError. Also, can't use ServerError until GODRIVER-1802 closes

Reply:
If checking for a single write error is actually an important use case, we need to reconsider the functionality of `ServerError`. Ideally, we want to get to the point where it's the main API for users to check server-side errors, so it seems weird for us not to use it here.

Also, I think GODRIVER-1802 can be merged because the associated PR has two LGTMs.